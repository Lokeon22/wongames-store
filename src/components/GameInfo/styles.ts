import styled, { css } from "styled-components"
import media from "styled-media-query"

import * as RibbonStyles from "../Ribbon/styles"
import * as HeadingStyles from "../Heading/styles"

export const Wrapper = styled.div`
  ${({ theme }) => css`
    position: relative;
    padding: ${theme.spacings.small};
    background-color: white;

    ${RibbonStyles.Wrapper} {
      right: -1.5rem;
      &:before {
        border-right-width: 1.5rem;
      }
    }

    ${media.lessThan("small")`
      ${HeadingStyles.Wrapper}{
        font-size: ${theme.font.sizes.xlarge};
      }
    `}

    ${media.greaterThan("medium")`
      ${RibbonStyles.Wrapper}{
        right: ${theme.spacings.small};
        top: ${theme.spacings.small};
        font-size: ${theme.font.sizes.large};
        &:before{
          border: none;
        }
      }
    `}
  `}
`

export const GameDescription = styled.h3`
  ${({ theme }) => css`
    color: ${theme.colors.gray};
    font-weight: normal;
    font-size: ${theme.font.sizes.small};
    margin-bottom: ${theme.spacings.small};

    ${media.greaterThan("medium")`
      max-width: 77rem;
    `}
  `}
`

export const ButtonsWrapper = styled.div`
  ${({ theme }) => css`
    display: flex;
    flex-direction: column;
    align-items: center;

    > button {
      width: 100%;
      margin-bottom: ${theme.spacings.small};
    }

    ${media.greaterThan("medium")`
      flex-direction: row-reverse;
      
      > button{
        width: initial;
        margin-bottom: 0;
      }
    `}
  `}
`
