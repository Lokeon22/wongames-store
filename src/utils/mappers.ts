import { BannersAttributes } from "../graphql/gqltypes/queryHomeBannersType"
import { GameAttributes } from "../graphql/gqltypes/queryGameType"
import { HighlightProps } from "../graphql/gqltypes/queryHighlightType"

export const bannerMapper = (banners: BannersAttributes[]) => {
  return banners.map(({ attributes }) => ({
    img: `http://localhost:1337${attributes.image.data.attributes.url}`,
    title: attributes.title,
    subtitle: attributes.subtitle,
    buttonLabel: attributes.button.label,
    buttonLink: attributes.button.link,
    ribbon: attributes.ribbon?.text || null,
    ribbonSize: attributes.ribbon?.size,
    ribbonColor: attributes.ribbon?.color
  }))
}

export const gamesMapper = (games: GameAttributes[]) => {
  return games.map(({ attributes }) => ({
    slug: attributes.slug,
    image: `http://localhost:1337${attributes.cover.data.attributes.url}`,
    title: attributes.name,
    developer: attributes.developers.data[0].attributes.name,
    price: attributes.price
  }))
}

export const highlightsMapper = (highlight: HighlightProps) => {
  return {
    title: highlight.title,
    subtitle: highlight.subtitle,
    backgroundImage: `http://localhost:1337${highlight.background.data.attributes.url}`,
    floatImage: `http://localhost:1337${highlight.floatImage.data.attributes.url}`,
    buttonLabel: highlight.buttonLabel,
    buttonLink: highlight.buttonLink,
    alignment: highlight.alignment
  }
}
