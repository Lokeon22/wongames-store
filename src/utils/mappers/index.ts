import { BannersAttributes } from "../../graphql/gqltypes/queryHomeBannersType"
import { GameAttributes } from "../../graphql/gqltypes/queryGameType"
import { HighlightProps } from "../../graphql/gqltypes/queryHighlightType"

import { formatPrice } from "../formatprice"
import { getImageUrl } from "../getImageUrl"

export const bannerMapper = (banners: BannersAttributes[]) => {
  return banners.map(({ attributes }) => ({
    img: `${getImageUrl(attributes.image.data.attributes.url)}`,
    title: attributes.title,
    subtitle: attributes.subtitle,
    buttonLabel: attributes.button.label,
    buttonLink: attributes.button.link,
    ribbon: attributes.ribbon?.text || null,
    ribbonSize: attributes.ribbon?.size,
    ribbonColor: attributes.ribbon?.color
  }))
}

export const gamesMapper = (games: GameAttributes[]) => {
  return games.map(({ attributes, id }) => ({
    id,
    slug: attributes.slug,
    image: `${getImageUrl(attributes.cover.data.attributes.url)}`,
    title: attributes.name,
    developer: attributes.developers.data[0].attributes.name,
    price: attributes.price
  }))
}

export const highlightsMapper = (highlight: HighlightProps) => {
  return {
    title: highlight.title,
    subtitle: highlight.subtitle,
    backgroundImage: `${getImageUrl(highlight.background.data.attributes.url)}`,
    floatImage: `${getImageUrl(highlight.floatImage.data.attributes.url)}`,
    buttonLabel: highlight.buttonLabel,
    buttonLink: highlight.buttonLink,
    alignment: highlight.alignment
  }
}

export const cartMapper = (cart: GameAttributes[] | undefined) => {
  return cart
    ? cart.map(({ id, attributes }) => ({
        id,
        slug: attributes.slug,
        title: attributes.name,
        price: formatPrice(attributes.price),
        img: `${getImageUrl(attributes.cover.data.attributes.url)}`
      }))
    : []
}
