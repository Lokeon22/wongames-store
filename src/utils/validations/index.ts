import Joi from "joi"

const fieldsValidations = {
  username: Joi.string().min(5).required(),
  email: Joi.string()
    .email({ tlds: { allow: false } })
    .required(),

  password: Joi.string().required(),
  confirm_password: Joi.string()
    .valid(Joi.ref("password"))
    .required()
    .messages({ "any.only": "confirm password does not match with password" })
}

export type FieldErros = {
  [key: string]: string
}

function getFieldErrors(objectError: Joi.ValidationResult) {
  const errors: FieldErros = {}

  if (objectError.error) {
    objectError.error.details.forEach((err) => {
      errors[err.path.join(".")] = err.message
    })
  }

  return errors
}

type signUpValues = {
  username: string
} & signInValues

export function signUpValidate(values: signUpValues) {
  const schema = Joi.object(fieldsValidations)

  return getFieldErrors(schema.validate(values, { abortEarly: false }))
}

type signInValues = {
  email: string
  password: string
}

export function signInValidate(values: signInValues) {
  const { email, password } = fieldsValidations

  const schema = Joi.object({ email, password })

  return getFieldErrors(schema.validate(values, { abortEarly: false }))
}

type forgotValidateParams = {
  email: string
}
export function forgotValidate(values: forgotValidateParams) {
  const { email } = fieldsValidations
  const schema = Joi.object({ email })

  return getFieldErrors(schema.validate(values, { abortEarly: false }))
}

type resetValidateParams = {
  password: string
  confirm_password: string
}
export function resetValidate(values: resetValidateParams) {
  const { password, confirm_password } = fieldsValidations
  const schema = Joi.object({ password, confirm_password })

  return getFieldErrors(schema.validate(values, { abortEarly: false }))
}
