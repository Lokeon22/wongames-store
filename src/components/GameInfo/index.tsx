"use client"
import { useSession } from "next-auth/react"
import { useWishlist } from "../../hooks/use-wishlist"
import {
  FavoriteBorder as FavIcon,
  Favorite
} from "@styled-icons/material-outlined"
import { formatPrice } from "../../utils/formatprice"

import * as S from "./styles"

import Heading from "../Heading"
import Ribbon from "../Ribbon"
import Button from "../Button"
import CartButton from "../CartButton"

export type GameInfoProps = {
  id: string
  title: string
  description: string
  price: number
  developer?: string
  slug?: string
  image?: string
}

const GameInfo = ({
  id,
  title,
  description,
  price,
  developer = "",
  image = "",
  slug = ""
}: GameInfoProps) => {
  const { data } = useSession()

  const { items, addToWishlist, removeFromWishlist } = useWishlist()
  const isWish = items.some((fav) => fav.id === id)

  const handleWish = () => {
    isWish
      ? removeFromWishlist(id)
      : addToWishlist({ id, title, price, developer, slug, image })
  }

  return (
    <S.Wrapper>
      <Heading lineBottom color="black">
        {title}
      </Heading>
      <Ribbon color="secondary">
        {price === 0 ? "FREE" : <>{formatPrice(price)}</>}
      </Ribbon>
      <S.GameDescription>{description}</S.GameDescription>
      <S.ButtonsWrapper>
        <CartButton id={id} size="large" hasText />
        {data && data.user && (
          <Button
            onClick={handleWish}
            icon={isWish ? <Favorite /> : <FavIcon />}
            minimal
          >
            Wishlist
          </Button>
        )}
      </S.ButtonsWrapper>
    </S.Wrapper>
  )
}

export default GameInfo
