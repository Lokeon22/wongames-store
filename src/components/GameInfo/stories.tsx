import { StoryFn, Meta } from "@storybook/react"

import GameInfo, { GameInfoProps } from "."
import { CartContextData } from "../../hooks/use-cart"
import items from "./mock"

export default {
  title: "GameInfo",
  component: GameInfo,
  args: items,
  parameters: {
    backgrounds: {
      default: "won-dark"
    }
  }
} as Meta

export const Default: StoryFn<GameInfoProps> = (args) => (
  <div style={{ maxWidth: "144rem", padding: "1.2rem" }}>
    <GameInfo {...args} />
  </div>
)

export const IsInCart: StoryFn<GameInfoProps & CartContextData> = (args) => (
  <div style={{ maxWidth: "144rem", padding: "1.2rem" }}>
    <GameInfo {...args} />
  </div>
)

IsInCart.args = {
  isInCart: () => true
}
