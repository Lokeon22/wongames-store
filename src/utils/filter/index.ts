import { ItemProps } from "../../components/ExploreSidebar"
import { ParsedUrlQueryInput } from "querystring"

type FilterItemProps = Pick<ItemProps, "name" | "type">

type ParseArgs = {
  queryString: ParsedUrlQueryInput
  filterItems: FilterItemProps[]
}

export const parseQueryStringToWhere = ({
  queryString,
  filterItems
}: ParseArgs) => {
  //eslint-disable-next-line @typescript-eslint/no-explicit-any
  const obj: any = {}

  Object.keys(queryString).forEach((key) => {
    const item = filterItems?.find((item) => item.name === key)
    const type = item?.type

    obj[key] = type !== "checkbox" ? queryString[key] : queryString[key]
  })

  return obj
}

//can create a object on if, example; :{any_name: queryString[key]}

export const parseQueryStringToFilter = ({
  queryString,
  filterItems
}: ParseArgs) => {
  //eslint-disable-next-line @typescript-eslint/no-explicit-any
  const obj: any = {}

  Object.keys(queryString).forEach((key) => {
    const item = filterItems?.find((item) => item.name === key)
    const isCheckbox = item?.type === "checkbox"
    const isArray = Array.isArray(queryString[key])

    obj[key] = !isArray && isCheckbox ? queryString[key] : queryString[key]
  })

  return obj
}

/*queryfilter example
{
  "limit": 15,
  "start": 0,
  "where_name": "",
  "where_price": 399,
  "where_category": "",
  "sort": "price:desc"
} */
