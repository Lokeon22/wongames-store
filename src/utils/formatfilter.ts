import { Values } from "../components/ExploreSidebar"

export const where_categories = [
  "Action",
  "Adventures",
  "Horror",
  "Fantasy",
  "Role-playing",
  "Jrpg",
  "Simulation",
  "Strategy",
  "Shooter"
]

export const categoryArrayFilter = (values: Values) => {
  return Object.keys(values)
    .map((key) => values[key])
    .filter(
      (value) =>
        typeof value === "string" &&
        !["price:asc", "price:desc"].includes(value)
    )
    .filter(Boolean)
}
