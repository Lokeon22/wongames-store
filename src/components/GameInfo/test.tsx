import { render, screen } from "../../utils/test-utils"

import GameInfo from "."
import items from "./mock"

jest.mock("next-auth/react", () => ({
  useSession: jest.fn(() => {
    return [{ session: null }]
  })
}))

describe("<GameInfo />", () => {
  it("should render the heading", () => {
    const { container } = render(<GameInfo {...items} />)

    expect(
      screen.getByRole("heading", { name: `${items.title}` })
    ).toBeInTheDocument()
    expect(
      screen.getByRole("heading", { name: `${items.description}` })
    ).toBeInTheDocument()
    expect(screen.getByText(/\$215.00/)).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it("should render the button", () => {
    render(<GameInfo {...items} />)

    expect(
      screen.getByRole("button", { name: /add to cart/i })
    ).toBeInTheDocument()
  })
})
