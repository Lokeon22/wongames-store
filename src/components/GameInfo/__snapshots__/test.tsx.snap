// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GameInfo /> should render the heading 1`] = `
.c8 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c4 {
  position: absolute;
  top: 1.6rem;
  display: flex;
  align-items: center;
  font-weight: 600;
  color: #FAFAFA;
  background-color: #3CD3C1;
  font-size: 1.4rem;
  padding: 0 2.4rem;
  height: 3.6rem;
  right: -2rem;
}

.c4::before {
  position: absolute;
  content: "";
  right: 0;
  border-style: solid;
  border-left-width: 0rem;
  border-right-color: transparent;
  border-bottom-color: transparent;
  border-bottom-width: 1rem;
}

.c4::before {
  border-left-color: #1f8a7d;
  border-top-color: #1f8a7d;
}

.c4::before {
  top: 3.6rem;
  border-top-width: 1rem;
  border-right-width: 2rem;
}

.c2 {
  font-weight: 600;
  color: #030517;
  position: relative;
  margin-bottom: 3.2rem;
  font-size: 2.8rem;
}

.c2::after {
  position: absolute;
  left: 0;
  bottom: -1rem;
  content: "";
  width: 5rem;
  border-bottom: 0.5rem solid #F231A5;
}

.c0 {
  position: relative;
  padding: 2.4rem;
  background-color: white;
}

.c0 .c3 {
  right: -1.5rem;
}

.c0 .c3:before {
  border-right-width: 1.5rem;
}

.c5 {
  color: #8F8F8F;
  font-weight: normal;
  font-size: 1.4rem;
  margin-bottom: 2.4rem;
}

.c6 {
  display: flex;
  flex-direction: column;
  align-items: center;
}

.c6 >button {
  width: 100%;
  margin-bottom: 2.4rem;
}

.c7 {
  display: inline-flex;
  align-items: center;
  justify-content: center;
  background: linear-gradient(189deg, #ff5f5f 0%, #f062c0 50%);
  border-radius: 0.4rem;
  border: 0;
  cursor: pointer;
  color: #FAFAFA;
  padding: 0.8rem;
  text-decoration: none;
  height: 5rem;
  font-size: 1.6rem;
  padding: 0.8rem 4.8rem;
}

.c7:hover {
  background: linear-gradient(180deg, #e35565 0%, #d958a6 50%);
}

.c7 svg {
  width: 1.5rem;
}

.c7 svg+span {
  padding: 0.8rem;
}

@media (min-width: 768px) {
  .c2 {
    font-size: 2.8rem;
  }
}

@media (max-width: 450px) {
  .c0 .c1 {
    font-size: 2.0rem;
  }
}

@media (min-width: 768px) {
  .c0 .c3 {
    right: 2.4rem;
    top: 2.4rem;
    font-size: 1.8rem;
  }

  .c0 .c3:before {
    border: none;
  }
}

@media (min-width: 768px) {
  .c5 {
    max-width: 77rem;
  }
}

@media (min-width: 768px) {
  .c6 {
    flex-direction: row-reverse;
  }

  .c6 >button {
    width: initial;
    margin-bottom: 0;
  }
}

<div
  class="c0"
>
  <h2
    class="c1 c2"
    color="black"
    linecolor="primary"
  >
    Borderlands 3
  </h2>
  <h3
    class="c3 c4"
    color="secondary"
  >
    $215.00
  </h3>
  <h3
    class="c5"
  >
    Experience the epic space strategy games that redefined the RTS genre. Control your fleet and build an armada across more than 30 single-player missions.
  </h3>
  <div
    class="c6"
  >
    <button
      class="c7"
    >
      <svg
        aria-hidden="true"
        aria-label="Add to cart"
        class="c8"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M11 9h2V6h3V4h-3V1h-2v3H8v2h3v3zm-4 9c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2zm10 0c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2zm-8.9-5h7.45c.75 0 1.41-.41 1.75-1.03l3.86-7.01L19.42 4l-3.87 7H8.53L4.27 2H1v2h2l3.6 7.59-1.35 2.44C4.52 15.37 5.48 17 7 17h12v-2H7l1.1-2z"
        />
      </svg>
      <span>
        Add to cart
      </span>
    </button>
  </div>
</div>
`;
